docs/python/notebooks/onnxruntime-nuphar-tutorial.ipynb

```diff
@@ -0,0 +1,681 @@
+    "    log_lines = f.readlines()\n",
+    "\n",
+    "log_lines[-10:]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The compiled code shows that the nodes of Add/Mul is fused in a single function, and vectorization is applied in the loop. This fusion happened inside compiler used by Nuphar execution provider, and it does not require model editing.\n",
+    "\n",
+    "Next, let's run inference on the model and compare with numpy in accuracy and speed:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'onnxruntime: 1.0402356000000001 seconds, numpy: 2.3407968 seconds'"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import onnxruntime\n",
+    "from onnx import numpy_helper\n",
+    "import numpy as np\n",
+    "from timeit import default_timer as timer\n",
+    "\n",
+    "seq = 128\n",
+    "batch = 16\n",
+    "input_data = np.random.rand(seq, batch, dim).astype(np.float32)\n",
+    "sess = onnxruntime.InferenceSession(simple_model_name)\n",
+    "feed = {X:input_data}\n",
+    "output = sess.run([], feed)\n",
+    "np_output = ((((input_data + input_data) * input_data) + input_data) * input_data) + input_data\n",
+    "assert np.allclose(output[0], np_output)\n",
+    "\n",
+    "repeats = 100\n",
+    "start_ort = timer()\n",
+    "for i in range(repeats):\n",
+    "    iter_start = timer()\n",
+    "    output = sess.run([], feed)\n",
+    "end_ort = timer()\n",
+    "start_np = timer()\n",
+    "for i in range(repeats):\n",
+    "    iter_start = timer()\n",
+    "    np_output = ((((input_data + input_data) * input_data) + input_data) * input_data) + input_data\n",
+    "end_np = timer()\n",
+    "'onnxruntime: {} seconds, numpy: {} seconds'.format(end_ort - start_ort, end_np - start_np)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "## 2. Create and run inference on a LSTM model\n",
+    "Now, let's take one step further to work on a 4-layer LSTM model, created from onnxruntime.nuphar.rnn_benchmark module."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "from onnxruntime.nuphar.rnn_benchmark import generate_model\n",
+    "lstm_model = 'LSTMx4.onnx'\n",
+    "input_dim = 256\n",
+    "hidden_dim = 1024\n",
+    "generate_model('lstm', input_dim, hidden_dim, bidirectional=False, layers=4, model_name=lstm_model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "**IMPORTANT**: since compiler generates code before data is fed, shape inference information is critical for optimization in compiler. To do that we run symbolic shape inference on the model."
```

Comment:
typos: 
"compiler" -> "the compiler"

"data" -> "the input data"

"optimization in compiler" -> "compiler optimizations"